security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,428 @@
+            this._activateSymetricIdentity(identityKey, (errActivate: Error) => {
+              if (errActivate) {
+                debug('Error from activate: ' + errActivate);
+                callback(errActivate);
+                this._fsm.transition('disconnected');
+              } else {
+                callback(null, null);
+              }
+            });
+          },
+        }
+      }
+    });
+  }
+
+  getEndorsementKey(callback: (err: Error, endorsementKey: string) => void): void {
+      this._fsm.handle('getEndorsementKey', callback);
+  }
+
+  getStorageRootKey(callback: (err: Error, storageKey: string) => void): void {
+    this._fsm.handle('getStorageRootKey', callback);
+  }
+
+  signWithIdentity(deviceIdData: Buffer, callback: (err: Error, signedResponse: Buffer) => void): void {
+    if (deviceIdData === null || deviceIdData.length === 0) {
+        throw new ReferenceError('\'deviceIdData\' cannot be \'' + deviceIdData + '\'');
+    }
+
+    if (this._idKeyPub == null) {
+        throw new errors.ArgumentError('activateIdentityKey first before signing');
+    }
+
+    this._fsm.handle('signWithIdentity', deviceIdData, callback);
   }
 
-  getStorageRootKey(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+  activateSymmetricIdentity(identityKey: Buffer, callback: (err: Error, returnedActivate: Buffer) => void): void {
+    this._fsm.handle('activateSymmetricIdentiy', identityKey, callback);
   }
 
-  signWithIdentity(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+  getRegistrationId(callback: (err: Error, registrationId: string) => void): void {
+    if (this._registrationId) {
+      callback(null, this._registrationId);
+    } else {
+      this.getEndorsementKey( function (endorsementError: Error, endorsementKey: string): void {
+        if (endorsementError) {
+          callback(endorsementError, null);
+        } else {
+          let hasher = crypto.createHash('sha256');
+          hasher.update(endorsementKey);
+          let hashedEndorsement: Buffer = hasher.digest();
+          let basedEndorsement: string = base32Encode(hashedEndorsement, 'RFC4648');
+          let loweredEndorsment: string = basedEndorsement.toLowerCase();
+          this._registrationId = loweredEndorsment.replace(/=/g, '');
+          callback(null, this._registrationId);
+        }
+      }.bind(this));
+    }
   }
 
-  activateSymmetricIdentity(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+  private _createPersistentPrimary(pki: PersKeyInfo, callback: (err: Error, resultPublicKey: TPMT_PUBLIC) => void): void {
+    this._tpm.allowErrors().ReadPublic(pki.handle, (resp: tss.ReadPublicResponse) => {
+      let rc = this._tpm.getLastResponseCode();
+      debug('ReadPublic(' + pki.name + ') returned ' + TPM_RC[rc] +  (rc === TPM_RC.SUCCESS ? '; PUB: ' + resp.outPublic.toString() : ''));
```

Comment:
>debug [](start = 6, length = 5)

generic remark: make sure nothing sensitive gets loged by debug() statements (i don't know if any of this is)

Reply:
In general, there is no "sensitive" data available.  Purpose of the tpm.  Will take care.  Thank you.

---
In reply to: [157890496](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r157890496) [](ancestors = 157890496)